src/sparseml/pytorch/sparsification/pruning/modifier_pruning_rigl.py

```diff
@@ -0,0 +1,599 @@
+        :param kwargs: optional kwargs to support specific arguments
+            for individual modifiers.
+        """
+        if (
+            "data_loader_builder" not in kwargs["grad_sampler"]
+            and "loss_function" not in kwargs["grad_sampler"]
+        ):
+            raise RuntimeError(
+                "grad_sampler dict with data_loader_builder and loss_function "
+                "must be provided to initialize GradSampler"
+            )
+
+        self._grad_sampler = GradSampler(
+            kwargs["grad_sampler"]["data_loader_builder"](**self._grad_sampler_kwargs),
+            kwargs["grad_sampler"]["loss_function"],
+        )
+
+        super().initialize(module, epoch, loggers, **kwargs)
+
+    def _get_scorer(self, params: List[Parameter]) -> PruningParamsGradScorer:
+        """
+        :param params: list of Parameters for scorer to track
+        :return: param scorer object to be used by this pruning algorithm
+        """
+        return RigLPruningParamsScorer(
+            params=params,
+            sparsity=self._final_sparsity,
+            sparsity_strategy=self._sparsity_strategy,
+        )
+
+    @staticmethod
+    def _reset_momentum_buffer(optimizer):
+        # multiply momentum buffer my param mask
+        if "state" in optimizer.state_dict():
+            for param_group in optimizer.param_groups:
+                for param in param_group["params"]:
+                    state = optimizer.state[param]
+                    param_mask = param.ne(0)
+                    if "momentum_buffer" in state:
+                        state["momentum_buffer"].mul_(param_mask)
+                    # Adam-specific additional buffers.
+                    if "exp_avg" in state:
+                        state["exp_avg"].mul_(param_mask)
+                    if "exp_avg_sq" in state:
+                        state["exp_avg"].mul_(param_mask)
+                    if "max_exp_avg_sq" in state:
+                        state["exp_avg"].mul_(param_mask)
+
+    def _get_update_fraction(self, epoch):
+        """
+        Returns the fraction of params updated at the current epoch.
+
+        :param epoch: current epoch
+        """
+        update_fraction = cosine_schedule(
+            epoch - self._start_epoch,
+            self._end_epoch - self._start_epoch,
+            self._init_update_fraction,
+            0.0,
+        )
+        _LOGGER.info(f"RigL mask Update fraction {update_fraction}")
+        return update_fraction
+
+    def check_mask_update(
+        self, module: Module, epoch: float, steps_per_epoch: int, **kwargs
+    ):
+        # enable grad buffering
+        torch.cuda.empty_cache()
+        if self._scorer._is_main_proc:
+            self._scorer._enabled_grad_buffering = True
+        # reset masks
+        for mask in self._module_masks._param_masks:
+            mask.data = torch.ones_like(mask)
+        # collect grad
+        self._collect_grad_samples(module, self._grad_sampler)
```

Comment:
If I understand this correctly we're accumulating the gradients between mask updates. What I understood from the paper is that this only done in the steps where the mask is updated (and thus avoid computing "dense" gradients every step).

I think that if we do it on during the mask update we can simplify the implementation since if fits the base class premise more closely

Reply:
If I understand correctly, this method only gets called when masks should be updated, (when update_ready is true). So we should not be sampling or accumulating any gradients during those steps when the masks are not updated.